### ISSUE TITLE
[MKL-DNN] Wrong API 

### ISSUE THREAD

**Issue by blueberry** [bytedeco/javacpp-presets#608]
This is a consequence of changes introduced in solving https://github.com/bytedeco/javacpp/issues/251

An example is the method `mkldnn_primitive_create`. Previously, there was only one version of this method, `mkldnn_primitive_create(mkldnn_primitive result, mkldnn_primitive_desc d, mkldnn_primitive_at_t sources, mkldnn_primitive destinations)`

The problem with the old method is that it did not support an array of primitives as destinations, so `PointerPointer` overload has been introduced: `mkldnn_primitive_create(PointerPointer result, mkldnn_primitive_desc d, mkldnn_primitive_at_t sources, PointerPointer destinations)`.

However, this new overload does not match the underlying C API of the original mkldnn, since it replaced **both** `mkldnn_primitive` arguments with `PointerPointer`. The original API requires `mkldnn_primitive` as the first argument, and `mkldnn_primitive` or `PointerPointer` as the last argument.

The (right) overload that is still missing should be:

`mkldnn_primitive_create(mkldnn_primitive result, mkldnn_primitive_desc d, mkldnn_primitive_at_t sources, PointerPointer destinations)`. 

Basically, this mismatch is present in all creation methods that were overloaded with this change. The intention of the API is that the `mkldnn_primitive` that is being initialized is always a single object, while other arguments sometimes must support `PointerPointer`.

**Comment by saudet** [comment]
The C API doesn't offer enough information to do that kind of mapping
automatically. We'd need to do a lot more manually:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#mapping-a-declaration-to-custom-code

Or we could provide the overloads as part of a helper class on top that is
less work to maintain than Info.javaText():
https://github.com/bytedeco/javacpp-presets/blob/master/opencv/src/main/java/org/bytedeco/javacpp/helper/opencv_core.java

**Comment by saudet** [comment]
Yes, the easiest thing would be to add that overload to the helper.

My changes have not removed anything, only adding new overloads, just not
exactly the ones that you're interested in.

**Comment by blueberry** [comment]
Yes, and some of the overloads that you added do not fit well with mkldnn's C api. At many places where `PointerPointer` has been added, it does not make sense for particular arguments, because they have cardinality `1` in the C API.

**Comment by saudet** [comment]
I told you, the C API doesn't have that information anywhere, so we have to
do it manually, and you want to help that's great, so I gave you material
to look at for that. Take a look at it and if you have any questions, go
ahead, ask away and I'll answer. Sounds good?

**Comment by blueberry** [comment]
Well, I have offered that help, read the guide you posted (haven't found the answers relevant to this issue there) and asked away with what I understand is the next steps. I need some handholding there. It is still not clear from your answers whether I got it right with that example that I posted :)

**Comment by saudet** [comment]
Yes, that's fine. Put it in the `helper` package along with a definition calling the native one that is generated but not exactly like you want.

The target class is generated yes, you can modify it if you want, but your changes will be overwritten by the parser. It's all documented here:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes
What should I be saying there to make that clear?

**Comment by blueberry** [comment]
OK, maybe I should ask more specific questions, that can be answered with yes/no or very brief explanations. The point here is that we are talking about one specific type of a simple change. That I can make without becoming expert in JavaCPP.

1. Is https://github.com/bytedeco/javacpp-presets/commit/6a8d6138dda7ba2141a389a036ae1efd2730f810 a native generated change, or something you had to write?
2. Can I make changes to that file directly, since the intention of changes is to make the Java API match the C api's semantic, by fixing the mismatch introduced in that commit?
3. In the case of the example I posted earlier,
```java
public static native @Cast("mkldnn_status_t") int mkldnn_stream_submit(mkldnn_stream stream,
        @Cast("size_t") long n, @Cast("mkldnn_primitive_t*") PointerPointer primitives,
        @ByPtrPtr mkldnn_primitive error_primitive);
```
why stream does not have @ByPtrPtr, or why error_primitive does (I am not seeking full understanding, just a few simple rules related to this particular example)?
4. Can you point me to a (sub)project that does a similar kind of custom mapping done via helper package? 

The point here is that, while I certainly do not have time and resources to became a JavaCPP expert, the changes that need to be done here are rather simple and straightforward, and I know the mkl-dnn C api enough to do the actual matching of each function. I am willing to do the boring part of work, but I need a skeleton to bootstrap that work.

**Comment by saudet** [comment]
1. Both
2. Yes, but your changes will be overwritten
3. Because that's how the function is defined in C
4. I gave you one, OpenCV

You can "bootstrap" from OpenCV, so again, take a look, if you have any questions, let me know. I understand you don't want to learn about JavaCPP, but if no one else does, I will remain the only one to make "small changes" and it's impossible for me to do everything on my own. This one in very low priority since it affects in no way the usability of the wrappers.

**Comment by blueberry** [comment]
Yes, I understand your perspective. Unfortunately, I currently do not have enough time to learn enough JavaCPP to bootstrap this on my own. I'll see to find some time in the future to get back to this...

**Comment by saudet** [comment]
Since it was missing, I've added a section about helper classes to the mapping recipes, that should help:
https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#writing-additional-code-in-a-helper-class
Let me know if there's anything unclear! Thanks

**Comment by blueberry** [comment]
Thank you. It probably does help. One thing that I think helps most with such thing is a tutorial-ish type of writing. Having detailed reference is of course essential, but to get to that point one has to have a good mental model how things work. At first, a trivial project with a walk-through post is what helps the most. I see that there is some writing with that in mind on the javacpp page, but there may be too much specific details right away, there is lots of text, but I can't find the actual project that one can clone and start experimenting with some brief hints.

On the other hand, the issue here is obviously my lack of time to get into this properly rather than insufficient documentation.

You're really doing a great work here! Thank you for the effort.

**Comment by saudet** [comment]
To start a presets from scratch, the sample project is here:
https://github.com/bytedeco/javacpp-presets/wiki/Create-New-Presets
Where are you expecting to find a link to this page?

**Comment by saudet** [comment]
Yeah, but then people get confused because in general end users shouldn't be reading them... For those that are actually interested in doing something I think there's now enough information that's easily reachable. It's just a lack of interest in working on such tools. It's not something people are interested in working on unfortunately.

**Comment by blueberry** [comment]
Hmmm. Generally, yes, but there stil *are* people at least potentially interestested. Sometimes it might be that it seems to them getting started is harder than it actually is...

**Comment by saudet** [comment]
Well, you know, if there were a lot of people interested, I'd be getting help with documentation, but it's not happening, so that's that :)

**Comment by blueberry** [comment]
I did not say *a lot*. I just said "more than zero" :)

**Comment by saudet** [comment]
Yeah, so there are ways to communicate with me (GitHub issues, Google Groups, private email, Gitter channels, Skype, direct phone calls, snail mail, etc) so I can guide them and stuff, try to divide the work, answer their questions, etc, but that's not happening either. People are just not interested in working together in making this tool better! Why do you deny this? Do you know of any other tools out there that come close to JavaCPP's capabilities? I'd like to know about them. I'd go work for the guy that creates something better than JavaCPP, but AFAIK that's not happening either.

**Comment by blueberry** [comment]
This list https://github.com/bytedeco/javacpp-presets/graphs/contributors says that 45 people helped in some way. Granted, most of these contributions might be trivial, but these people were interested in this project at least to some extent and helped a bit...

**Comment by saudet** [comment]
Yes, there are some people like you that help, sure, and I'm very grateful. There are a lot of happy end users, so I'm happy to keep doing most of the work myself until someone else at Oracle or Google or wherever with some capital realizes the importance of this, or not. What's your point? If you'd like to help in some other ways, go for it, but you said yourself you don't have time. So what exactly are you trying to tell me here? You don't have the time, I don't have the time, so who exactly should be doing all that you want me to do?

**Comment by saudet** [comment]
I think it's safe to assume that the end users of the presets do not read the README.md file of JavaCPP itself, so I've added a "Getting Started" section there to emphasize the new guide:
https://github.com/bytedeco/javacpp#getting-started

**Comment by saudet** [comment]
I've also added this today, that should help too:
https://github.com/bytedeco/javacpp/wiki/Basic-Architecture
What do you think?

**Comment by blueberry** [comment]
Thanks. Looks very informative and to the point. I am working on some other
things right now but as soon as I return to mkldnn I’ll see if I can help
with specific questions and clarifications.

**Comment by saudet** [comment]
To map things like this more easily, I'm thinking we could augment `Info` to take into account the name of arguments, etc, such that we could do something like this:
```java
new Info("mkldnn_primitive_t* error_primitive").pointerTypes("@ByPtrPtr mkldnn_primitive"))
```

BTW, I've written here a blog post about the importance of this work:
http://bytedeco.org/news/2019/01/11/importance-of-a-distribution/
Please forward to potentially interested parties :)
